rand_distr/src/dirichlet.rs

```diff
@@ -93,36 +93,33 @@ where
             }
         }
 
-        Ok(Dirichlet { alpha: alpha.to_vec().into_boxed_slice() })
+        Ok(Dirichlet { alpha })
     }
 
     /// Construct a new `Dirichlet` with the given shape parameter `alpha` and `size`.
     ///
     /// Requires `size >= 2`.
     #[inline]
-    pub fn new_with_size(alpha: F, size: usize) -> Result<Dirichlet<F>, Error> {
+    pub fn new_with_size(alpha: F) -> Result<Dirichlet<F, N>, Error> {
         if !(alpha > F::zero()) {
             return Err(Error::AlphaTooSmall);
         }
-        if size < 2 {
+        if N < 2 {
             return Err(Error::SizeTooSmall);
         }
-        Ok(Dirichlet {
-            alpha: vec![alpha; size].into_boxed_slice(),
-        })
+        Ok(Dirichlet { alpha: [alpha; N] })
     }
 }
 
-impl<F> Distribution<Vec<F>> for Dirichlet<F>
+impl<F, const N: usize> Distribution<[F; N]> for Dirichlet<F, N>
 where
     F: Float,
     StandardNormal: Distribution<F>,
     Exp1: Distribution<F>,
     Open01: Distribution<F>,
 {
-    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec<F> {
-        let n = self.alpha.len();
-        let mut samples = vec![F::zero(); n];
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> [F; N] {
+        let mut samples = [F::zero(); N];
```

Comment:
Nice not to have to allocate here. Also nice that the output length is fixed.

Given `Distribution::sample`'s API, storage can't be an input. A possible alternative would be to use a generic associative type (`type Result<'s, T> = &'s T;` for `Distribution<T>`)... yes, that's horrible!

Reply:
I am sorry but I don't understand this comment. What does it mean that storage can't be an input?

Reply:
No worries, it's not really the topic of this PR.

Below the storage (`buf`) is an input to `sample`, but we can't do this without changing the trait (in ways that don't make sense for most other uses):
```rust
let mut rng = ...;
let dist = Dirichlet::new(..)?;
let mut buf = [0; 3];
dist.sample(&mut rng, &mut buf);
```